FormalSystems/Chomsky/Regular/Nerode.lean

```diff
@@ -24,9 +54,65 @@ theorem Nerode.not_right_congruent (L : Language α) (u v : Word α) (pre: u ≠
     have uw_eq_vw := e ▸ h (u*w) ⟨ Or.inl rfl, uw_in_l ⟩
     exact pre <| Word.mul_right_cancel uw_eq_vw
 
-theorem Nerode.not_right_congruent1 (L : Language α) (u v : Word α):
+theorem MyhillNerodeRelation.not_congruent_if':
   (∃w, u*w ∈ L ∧ v*w ∉ L) → ¬ u ≃(L) v := by
   intro ⟨ w, uw_in_l, vw_nin_l ⟩ contra
   apply vw_nin_l
   apply (contra w).mp
   exact uw_in_l
+
+theorem MyhillNerodeRelation.mem_language:
+  u ≃(L) v → (u ∈ L ↔ v ∈ L) := fun h => by
+    have := h ε
+    simp at this
+    assumption
+
+theorem MyhillNerodeRelation.mem_language':
+  u ≃(L) v → ((u ∈ L) = (v ∈ L)) :=
+    propext ∘ MyhillNerodeRelation.mem_language
+
+def DecisionProcedure (L: Language α): Type := DecidablePred (· ∈ L)
+
+-- classically, every Language has a decision procedure, so we might choose one
+noncomputable def DecisionProcedure.classical: DecisionProcedure L := fun w =>
+  Classical.choice $
+    match Classical.em (w ∈ L) with
+    | .inl h => Nonempty.intro $ Decidable.isTrue h
+    | .inr h => Nonempty.intro $ Decidable.isFalse h
```

Comment:
It also works to use [`Classical.decPred`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Basic.html#Classical.decPred) directly.
```suggestion
noncomputable def DecisionProcedure.classical: DecisionProcedure L := Classical.decPred (· ∈ L)
```